File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(1000000)
def dfs(x, y):
    global r
    graph[x][y] += 1
    r += 1
    for d in range(4):
        nx = x + dx[d]
        ny = y + dy[d]
        if 0 <= nx < n  and 0 <= ny < n:
            if graph[nx][ny] == 1:
                dfs(nx, ny) 

n = int(input())

graph = [list(map(int, list(input().rstrip()))) for _ in range(n)]

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

cnt = 0
# r = 0
result = []
for i in range(n):
    for j in range(n):
        if graph[i][j] == 1:
            r = 0
            dfs(i, j)
            result.append(r)
            cnt += 1
result.sort()
print(cnt, *result, sep='\n')

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Main {
    static List<String> graph;
    static boolean[][] visited;
    static BufferedReader br;
    static int cnt, dCnt, N;
    static final int[] dx = {0, 0, -1, 1}, dy = {-1, 1, 0, 0};
    static List<Integer> result;
    static StringBuilder sb;

    public static void main(String[] args) throws IOException {
        br = new BufferedReader(new InputStreamReader(System.in));
        N = readInt();
        graph = new ArrayList<>();
        visited = new boolean[N][N];
        result = new ArrayList<>();

        for (int i = 0; i < N; i++) {
            graph.add(br.readLine());
        }
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                dCnt = 0;
                char c = graph.get(i).charAt(j);
                if (c == '1' && !visited[i][j]) {
                    cnt++;
                    dfs(i, j);
                    result.add(dCnt);
                }
            }
        }

        Collections.sort(result);
        sb = new StringBuilder();
        sb.append(cnt).append('\n');
        for (int i : result) {
            sb.append(i).append('\n');
        }
        System.out.println(sb);
    }

    private static void dfs(int r, int c) {
        visited[r][c] = true;
        dCnt++;
        for (int d = 0; d < 4; d++) {
            int y = r + dy[d];
            int x = c + dx[d];
            if (y < 0 || y >= N || x < 0 || x >= N) {
                continue;
            }
            char a = graph.get(y).charAt(x);
            if (a == '1' && !visited[y][x]) {
                dfs(y, x);
            }
        }
    }

    private static int readInt() throws IOException {
        int v, t = 0;
        while ((v = System.in.read()) > 47) {
            t = t * 10 + (v - 48);
        }
        if (v == 13) {
            System.in.read();
        }
        return t;
    }
}